Userland/Libraries/LibCore/Stream.h

```diff
@@ -0,0 +1,846 @@
+    }
+    bool is_success() const { return !is_error(); }
+
+    bool is_kresult() { return m_value.has<KResult>(); }
+    bool is_getaddrinfo_error() { return m_value.has<GetAddrInfoError>(); }
+
+    KResult as_kresult() { return m_value.get<KResult>(); }
+    GetAddrInfoError as_getaddrinfo_error()
+    {
+        return m_value.get<GetAddrInfoError>();
+    }
+
+    StringView getaddrinfo_error_string()
+    {
+        VERIFY(is_getaddrinfo_error());
+        return { gai_strerror(static_cast<int>(as_getaddrinfo_error())) };
+    }
+
+private:
+    Variant<KResult, GetAddrInfoError> m_value;
+};
+
+/**
+ * The Socket class is the base class for all concrete BSD-style socket classes.
+ * Sockets are non-seekable streams which can be read byte-wise.
+ */
+class Socket : public Stream {
+public:
+    /// Checks how many bytes of data are currently available to read on the
+    /// socket. For datagram-based socket, this is the size of the first
+    /// datagram that can be read. Returns either the amount of bytes, or an
+    /// errno in the case of failure.
+    virtual KResultOr<size_t> pending_bytes() const = 0;
+    /// Returns whether there's any data that can be immediately read, or an
+    /// errno on failure.
+    virtual KResultOr<bool> can_read_without_blocking(int timeout = 0) const = 0;
+
+    Socket(Socket&&) = default;
+    Socket& operator=(Socket&&) = default;
+
+    // Sets the blocking mode of the socket. If blocking mode is disabled, reads
+    // will fail with EAGAIN when there's no data available to read, and writes
+    // will fail with EAGAIN when the data cannot be written without blocking
+    // (due to the send buffer being full, for example).
+    virtual KResult set_blocking(bool enabled) = 0;
+    // Sets the close-on-exec mode of the socket. If close-on-exec mode is
+    // enabled, then the socket will be automatically closed by the kernel when
+    // an exec call happens.
+    virtual KResult set_close_on_exec(bool enabled) = 0;
+
+    Function<void()> on_ready_to_read;
+
+protected:
+    enum class SocketDomain {
+        Local,
+        Inet,
+    };
+
+    enum class SocketType {
+        Stream,
+        Datagram,
+    };
+
+    Socket()
+    {
+    }
+
+    static KResultOr<int> create_socket(SocketDomain, SocketType);
+    // FIXME: This will need to be updated when IPv6 socket arrives. Perhaps a
+    //        base class for all address types is appropriate.
+    static Result<IPv4Address, SocketError> resolve_host(String const&, SocketType);
+
+    static KResult connect_local(int fd, String const& path);
+    static KResult connect_inet(int fd, SocketAddress const&);
+};
+
+// Concrete classes.
+
+enum class OpenMode : unsigned {
+    NotOpen = 0,
+    Read = 1,
+    Write = 2,
+    ReadWrite = 3,
+    Append = 4,
+    Truncate = 8,
+    MustBeNew = 16,
+    KeepOnExec = 32,
+    Nonblocking = 64,
+};
+
+AK_ENUM_BITWISE_OPERATORS(OpenMode)
+
+class File final : public SeekableStream {
```

Comment:
I'd split these to multiple files, but I think that can be done when we've gotten rid of `IODevice` so `File.h` and co are available.

Reply:
Yep, that's the intention here.